### ISSUE TITLE
Income through Bolt Generation + Scavenging mechanic

### ISSUE THREAD

**Issue by luttje** [luttje/gmod-experiment-redux#10]
Bolts are the currency of Experiment and a big part of the game is to increase wealth by hoarding them. In classic Experiment the most effective way to gain bolts was by building Bolt Generators, hiding them and protecting them. Over time they generate income, only requiring the player to charge them every half-hour or so by simply interacting with them. The biggest risk involved was someone breaking into their hiding place and destroying the generators for the bolt reward that gave.

This had the obvious downside that players would have to hide inside and protect their generators during their first days of playing.

Although this 'AFK grinding' had its appeal, its not something I think will be very fun nowadays. Besides I had some ideas on how to keep gameplay diverse and interesting.

An additional technical issue was that Bolt Generators were physics objects. In classic Experiment the many generators could start lagging the server over time.

## 🏭 Bolt Control Units

To combat the server lag due to many physics objects, we invented the BCU (Bolt Control Unit). Its a generator that can be spawned where the player is looking, and picked up into inventory, but not moved around. Its stationary and immovable when it has been placed into the world.

Instead of spawning many different kinds of bolt generators (like in Classic Experiment) BCU's can be upgrades. Higher upgrades require a perk. Higher upgrades result in more bolt output.

Like in Classic Experiment BCU's have to be charged in order to output bolts.

In Classic Experiment bolt generates sent bolts 'magically' to the player every 10 minutes. Late in the last iteration of Experiment we considered and tested the idea of the BCU holding on to generated bolts. Requiring the player to extract them from the device manually. This caused for the following fun: 
> An alliance has a giant base filled with tons of generators. They're overrun by a rival alliance who take control of the base, including generators.
> The rival alliance is now keeping the generators 'hostage', extracting bolts from it for themselves.

> Old image from Revolt I found somewhere in my backups
> ![thebcu](https://github.com/luttje/gmod-experiment-redux/assets/2738114/2642f8cd-e4bd-48ce-89c4-66918bceea33)

## 🔋 Powering Bolt Control Units through Scavenging

To generate bolts, bolt generators like the BCU required Power. Power was free and invisible. The generator would say 'Power: 0', you'd just press 'E' to interact with it (charging it to 'Power: 3') and it would be good to generate bolts for another half hour. This encouraged staying close to the generator. Players sometimes cooperated where they would charge each others generators in alternating shifts. The AFK players would stand idle in the corner.

Personally I find the most fun experiences in Experiment happened out in the world. Not hiding inside, camping to defend bolt generators (although that too had its appeal). In order to promote going outside I'm considering the following ideas:

* Bolt Control Units require 'Scrap' to increase Power
* Players fill the BCU with scrap by interacting with it
* Unlike Power in Classic Experiment, Scrap isn't free, players find it in the city
* Throughout the city are piles of junk, trashcans, dumpsters and other intractable containers, filled with materials that can be turned into scrap
* When a player fiends junk (e.g: empty can/metal shard/iron plate) they can interact with it in their inventory to turn it into 'Scrap'

This hopefully encourages players to go outside, run around and hoard scrap for themselves. Additionally we could hide other fun items alongside, to scavenging a new gameplay mechanic. Such items could be: single bullets (free ammo), tattered clothing (free low tier armor), diaries/books/paper (lore items to collect and learn the backstory of Experiment).

## TL;DR / Checklist

This TL;DR also functions as a checklist for the proposed items. Check out the current version of the project to experience:

- [x] Static (immovable) bolt generator type: Bolt Control Unit (BCU)
- [x] BCU generates bolts every 10 minutes
- [x] BCU can be upgraded to output more bolts
- [x] BCU hold on to generated bolts, players must extract it from the device manually
- [x] BCU requires power to function
- [x] Power can be increased by loading Scrap into BCU
- [x] Scrap can be found by scavenging in the city, exploring corners, trashcans, dumpsters and other intractable entities

Alongside scrap the player can find free items like:
- [x] lore items
- [x] single bullets
- [x] tattered clothing 

Personally I like this idea and I think it can enhance the experience, without changing too much of the core gameplay. 

I'd be interested to hear what others think

**Comment by luttje** [comment]
A demo:

https://github.com/luttje/gmod-experiment-redux/assets/2738114/809a1971-0a5d-40a4-889c-8705ef86ced3

**Comment by luttje** [comment]
> When you happen upon lore items while scavenging, you might learn something about the world you're in. A character in the city may also be interested in collecting such items.

**A demo of the NPC/mission system:**
https://github.com/luttje/gmod-experiment-redux/assets/2738114/b43835b9-961e-4e7d-a7be-e8e1d24eb05d

See this code for the logic and dialogue config: 
https://github.com/luttje/gmod-experiment-redux/blob/268decb14d79a008ead61a47149e78d48ea9676d/plugins/lore/npcs/sh_archivist.lua

**Comment by utaysi** [comment]
> Bolt Control Units require 'Scrap' to increase Power

I really like this idea. It would really make the streets of the city more active and force new combat on the streets as well. Like running the streets for compounds & scrap in Rust. 

- I also think the scraps should be tradeable items. This could open blackmarket dynamics between the players which could be interesting.

**Comment by luttje** [comment]
Good call, I'll keep scraps as "just items" without any further restrictions. You can trade/drop them, collect them for your alliance, or do whatever. Which keeps the door open for fun dynamics like you said.
